File: storage/storage_test.go
package storage

import (
	"os"
	"testing"
)

func TestStorageOperations(t *testing.T) {
	tempDir, err := os.MkdirTemp("", "note-tui-test-*")
	if err != nil {
		t.Fatalf("failed to create temp dir: %v", err)
	}
	defer os.RemoveAll(tempDir)

	store, err := NewStorage(tempDir)
	if err != nil {
		t.Fatalf("failed to init storage: %v", err)
	}

	// Test 1: Create Note
	note1, err := store.CreateNote("Meeting Notes", "# Team Sync\nDiscuss Q3 roadmap and feature priorities.")
	if err != nil {
		t.Fatalf("failed to create note: %v", err)
	}

	if note1.Filename != "Meeting-Notes.md" {
		t.Errorf("expected filename 'Meeting-Notes.md', got '%s'", note1.Filename)
	}

	if note1.Title != "Team Sync" {
		t.Errorf("expected title 'Team Sync', got '%s'", note1.Title)
	}

	if note1.WordCount != 9 {
		t.Errorf("expected 9 words, got %d", note1.WordCount)
	}

	// Test 2: Create Second Note
	_, err = store.CreateNote("Shopping List.txt", "1. Milk\n2. Coffee\n3. Bread")
	if err != nil {
		t.Fatalf("failed to create second note: %v", err)
	}

	// Test 3: List Notes
	notes, err := store.ListNotes()
	if err != nil {
		t.Fatalf("failed to list notes: %v", err)
	}

	if len(notes) != 2 {
		t.Fatalf("expected 2 notes, got %d", len(notes))
	}

	// Test 4: Save / Update Note
	updatedNote, err := store.SaveNote(note1.Filename, "# Team Sync\nUpdated content with more details.")
	if err != nil {
		t.Fatalf("failed to update note: %v", err)
	}

	if updatedNote.WordCount != 8 {
		t.Errorf("expected 8 words in updated note, got %d", updatedNote.WordCount)
	}

	// Test 5: Filter Notes
	filtered := FilterNotes(notes, "Shopping")
	if len(filtered) != 1 {
		t.Fatalf("expected 1 match for 'Shopping', got %d", len(filtered))
	}
	if filtered[0].Title != "Shopping List" {
		t.Errorf("expected title 'Shopping List', got '%s'", filtered[0].Title)
	}

	// Test 6: Delete Note
	err = store.DeleteNote(note1.Filename)
	if err != nil {
		t.Fatalf("failed to delete note: %v", err)
	}

	notesAfterDelete, _ := store.ListNotes()
	if len(notesAfterDelete) != 1 {
		t.Fatalf("expected 1 note after deletion, got %d", len(notesAfterDelete))
	}
}

func TestCleanFilename(t *testing.T) {
	tests := []struct {
		input    string
		expected string
	}{
		{"my note", "my-note.md"},
		{"important/work:note*", "important-work-note-.md"},
		{"test.txt", "test.txt"},
		{"", "untitled.md"},
	}

	for _, tt := range tests {
		got := CleanFilename(tt.input)
		if got != tt.expected {
			t.Errorf("CleanFilename(%q) = %q; want %q", tt.input, got, tt.expected)
		}
	}
}

File: tests/storage.test.js
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { Storage, cleanFilename, filterNotes } from '../src/storage.js';

let tempDir;

beforeEach(async () => {
	tempDir = await fs.promises.mkdtemp(path.join(os.tmpdir(), 'note-tui-test-'));
});

afterEach(async () => {
	await fs.promises.rm(tempDir, { recursive: true, force: true });
});

describe('Storage operations', () => {
	it('creates, lists, saves, filters and deletes notes', async () => {
		const store = await new Storage(tempDir).init();

		const note1 = await store.createNote('Meeting Notes', '# Team Sync\nDiscuss Q3 roadmap and feature priorities.');
		expect(note1.filename).toBe('Meeting-Notes.md');
		expect(note1.title).toBe('Team Sync');
		expect(note1.wordCount).toBe(9);

		await store.createNote('Shopping List.txt', '1. Milk\n2. Coffee\n3. Bread');

		const notes = await store.listNotes();
		expect(notes).toHaveLength(2);

		const updated = await store.saveNote(note1.filename, '# Team Sync\nUpdated content with more details.');
		expect(updated.wordCount).toBe(8);

		const filtered = filterNotes(notes, 'Shopping');
		expect(filtered).toHaveLength(1);
		expect(filtered[0].title).toBe('Shopping List');

		await store.deleteNote(note1.filename);
		const afterDelete = await store.listNotes();
		expect(afterDelete).toHaveLength(1);
	});

	it('creates a note with a suffix when the file already exists', async () => {
		const store = await new Storage(tempDir).init();
		await store.createNote('Note', 'first');
		const second = await store.createNote('Note', 'second');
		expect(second.filename).toBe('Note-1.md');
		const notes = await store.listNotes();
		expect(notes).toHaveLength(2);
	});

	it('sorts notes by modification time, newest first', async () => {
		const store = await new Storage(tempDir).init();
		await store.createNote('Old', 'old content');
		const newer = await store.createNote('New', 'new content');
		await new Promise((resolve) => setTimeout(resolve, 20));
		await store.saveNote(newer.filename, 'edited newer content');

		const notes = await store.listNotes();
		expect(notes[0].filename).toBe(newer.filename);
	});

	it('ignores non-text files when listing', async () => {
		const store = await new Storage(tempDir).init();
		await fs.promises.writeFile(path.join(tempDir, 'ignore.me'), 'x');
		await store.createNote('Real', 'content');
		const notes = await store.listNotes();
		expect(notes).toHaveLength(1);
	});

	it('creates the storage directory automatically', async () => {
		const nested = path.join(tempDir, 'a', 'b');
		const store = await new Storage(nested).init();
		await store.createNote('Nested', 'hi');
		expect(fs.existsSync(path.join(nested, 'Nested.md'))).toBe(true);
	});

	it('renames a note', async () => {
		const store = await new Storage(tempDir).init();
		await store.createNote('Original', 'body');
		await store.renameNote('Original.md', 'Renamed');
		const notes = await store.listNotes();
		expect(notes[0].filename).toBe('Renamed.md');
	});
});

describe('cleanFilename', () => {
	it.each([
		['my note', 'my-note.md'],
		['important/work:note*', 'important-work-note-.md'],
		['test.txt', 'test.txt'],
		['', 'untitled.md'],
	])('cleanFilename(%j) = %j', (input, expected) => {
		expect(cleanFilename(input)).toBe(expected);
	});
});
